add open62541 recipe for v1.0 and v1.0.1

saukijan opened the pull request:
##  open62541/1.0.1

### Description
This PR adds a recipe for [open62541](https://github.com/open62541/open62541) project versions 1.0 and 1.0.1. 

Both version were tested under Linux Debian 10 gcc8 setup

Signed-off-by: Dovydas Girdvainis <dovydas.girdvainis@hahn-schickard.de>

Specify library name and version: **open62541/1.0.1**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=1687) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=1687) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=1687) it.</sub>

uilianries commented on recipes/open62541/all/patches/set_static_version_v_1_0_1.patch:
```diff
@@ -0,0 +1,589 @@
+-        UA_STRING_ALLOC(VERSION(UA_OPEN62541_VER_MAJOR, UA_OPEN62541_VER_MINOR,
+-                                UA_OPEN62541_VER_PATCH, UA_OPEN62541_VER_LABEL));
+-    #ifdef UA_PACK_DEBIAN
++    conf->buildInfo.softwareVersion = UA_STRING_ALLOC("1.0.1");
```
I see only few changes on these patches, I think they could be reduced.

saukijan replied:
I merged both versions of `set_static_version_*.patch` into a single one, however I dont see how I could do the same for `add_conan_center_index_compliance_*.patch` since root **CMakeLists.txt** was slightly  changed in between the two versions. I expermented with one mega patch for both versions, but it always ended up with `git: patch does not apply` error.

Croydon commented:
Maybe having a look at our Bincrafters' recipe could help 😄  

https://github.com/bincrafters/conan-open62541

saukijan commented:
> 
> 
> Maybe having a look at our Bincrafters' recipe could help 😄
> 
> https://github.com/bincrafters/conan-open62541

Well there are more open62541 recipes out there as well, however none of them offer full Open62541 CMake option support, further more the recipe you linked does not have full OPC UA node conformance (the archives it is linking to do NOT have the required submodules initialized, thus the missing OPC UA node definitions) nor will it work for CLang Debug configurations. This recipe addresses all of these concerns and hopefully is a bit more up to date with open62541 proejcet

saukijan commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,234 @@
+    }
+    default_options = {
+        "fPIC": True,
+        "shared": True,
```
Thanks for checking. I caused this option to no longer work by removing the version setting utility from the cmake. One of the variables set by this utility was being used by the amalgamator to generate the single header file. I fixed it by setting the required variable in my `set_static_version.patch`. 

I will re-run local tests for all of the options again to make sure that my latest patches doesn't break anything else. I will push this change once my tests pass.

madebr replied:
shared should be False by default
See @uilianries comment

SSE4 replied:
please check shared to False by default and apply the suggestion.
it's important as static build is safe default - not all platforms support shared libraries.

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+        "discovery_semaphore": True,
+        "discovery_multicast": False,
+        "query": False,
+        "encription": False,
```
This should be encryption

saukijan replied:
Resolved in the next commit, just running local tests

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+
+    def requirements(self):
+        if self.options.encription:
+            self.requires("mbedtls/2.16.3-gpl")
```
Are other encryption libraries possible? Like openssl?
If so, I'd change the option encryption to receive a list of possible encryption packages

syoliver replied:
openssl is possible with version 1.1 (see my pr).
With mbedtls, why forcing gpl variant ?

saukijan replied:
This specific commit is focusing on 1.0 and 1.0.1 versions which do not support OpenSSL as an option, I started this PR before 1.1 was even released as a Release Candidate and want to finish that before adding support for v1.1 (which won't be that hard once this is done, and yes I do plan to add OpenSSL as an option) 

Regarding [mbedtls](https://github.com/conan-io/conan-center-index/blob/master/recipes/mbedtls/all/conanfile.py) and GPL-2.0 or Apache-2.0 - I am not aware of any difference in  code apart of the different Licenses and frankly I do not understand the need for two different licenses, I simply chose the one I had installed on my machine. I could provide an option for the user to choose one over the other, but does it change anything significant in the code base, which would warrant the added option? If not I would stay on the side of simplicity and skip it.

danimtb replied:
I'd prefer to go for the less restrictive license here

saukijan replied:
I'll change it to Apache then

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+        if self.options.single_header:
+            self.cpp_info.defines.append("UA_ENABLE_AMALGAMATION")
+        if self.settings.os == "Windows":
+            self.cpp_info.libs.append("ws2_32")
```
system_libs

saukijan replied:
Resolved in the next commit, just running local tests

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("LICENSE-CC0", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
```
Did you remove the cmake scripts?

saukijan replied:
What do you mean by CMake Scripts? I removed the custom packaging and versioning Cmake modules with a [compatibility patch for v1.0](https://github.com/Hahn-Schickard/conan-center-index/blob/master/recipes/open62541/all/patches/add_conan_center_index_compliance_v_1_0.patch) and [compatibility patch for v1.0.1](https://github.com/Hahn-Schickard/conan-center-index/blob/master/recipes/open62541/all/patches/add_conan_center_index_compliance_v_1_0_1.patch) that open62541 team developed. 

If you are interested here is a list of changes both patches apply: 

* remove git based versioning Cmake module: [tools/cmake/SetGitBasedVersion.cmake](https://github.com/Hahn-Schickard/open62541/blob/master/tools/cmake/SetGitBasedVersion.cmake)
* remove AddSan usage for Clang Debug setups
* remove open62541.pc.in file usage and open62541.pc file generation
* remove open62541_install_tools and open62541_install_nodeset (removes nodeset generation, which requires various shell scripts)
* remove open62541Config.cmake and open62541ConfigVersion.cmake installation
* remove open62541_install_examples
* remove deb [packaging utilities](https://github.com/Hahn-Schickard/open62541/blob/master/tools/packaging/CMakeLists.txt)

madebr replied:
Oh, I didn't check the patches. Thanks for explaining.

In my rcipes, I remove these files in the conan script instead of patching the sources.
This scales better for newer releases. (For now)

madebr commented on recipes/open62541/all/conandata.yml:
```diff
@@ -0,0 +1,18 @@
+sources:
+  "1.0.0":
```
1.0.0 can be removed, I think.
Nobody will use an outdated version

syoliver replied:
Last version is 1.1 anyway

saukijan replied:
There are people who use older versions due to compatibility reasons, and it does not cause problems to keep it as an option

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+        "pub_sub": ["None", "Simple", "Ethernet", "Ethernet_XDP"],
+        "data_access": [True, False],
+        "compiled_nodeset_descriptions": [True, False],
+        "namescpae_zero": ["MINIMAL", "REDUCED", "FULL"],
```
namespace_zero typo

saukijan replied:
Resolved in the next commit, just running local tests

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+        if Version(self.version) <= "1.0.1":
+            self._download_ua_nodes()
+            self._download_mdnsd()
+        self._patch_sources()
```
Please patch in build

saukijan replied:
Resolved in the next commit, just running local tests

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+
+    def _download_ua_nodes(self):
+        link = ""
+        if self.version == "1.0.0":
```
Please use Conan packages for the dependencies.
Or add these extra zips to the conandata.yml file.

syoliver replied:
nodeset is not a c++ dependency, just a submodule. I have the same issue in my PR, where I added a yml for description of submodules by version.

saukijan replied:
This is impossible due to the way github archives projects with sub-module repositories. I opened an issue at open62541 about it, but it seems they can not resolve it.

I could set up separate recipes for the [UA Node set ](https://github.com/OPCFoundation/UA-Nodeset/tree/0777abd1bc407b4dbd79abc515864f8c3ce6812b) and [mdsnd](https://github.com/Pro/mdnsd/tree/4bd993e0fdd06d54c8fd0b8f416cda6a8db18585) repositories, however they would have to be accepted to make this PR work.

saukijan replied:
It seems ua node set might cause packaging problems due to licensing. I will inquirer by open62541 for more info

Pro replied:
There is already a discussion going on, on how to handle the nodeset submodule.
See: https://github.com/open62541/open62541/issues/3710

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+    def _download_mdnsd(self):
+        link = ""
+        if self.version == "1.0.0":
+            link = "https://github.com/Pro/mdnsd/archive/8fe3a7e7e9d0a9196b126c64f0d1905569b83d40.zip"
```
Likewise. Use Conan packages or add to conandata.yml.

saukijan replied:
[mdnsd](https://github.com/Pro/mdnsd] is not a Conan dependency

madebr replied:
Having all dependencies as conan recipes/packages would be the correct way...
I don't think the maintainers will accept this recipe if it fetches its dependencies as git submodules.
mdnsd might be useful for somebody else..

saukijan replied:
alright, I'll develop those recipes as well, but until they are accepted and merged in, I'll keep this method, so I can continue to test on the central CI/CD pipeline. I'll upload a git diff for this method as well as links to the new recipes in this conversation.

madebr replied:
Great!

saukijan replied:
As a first step, I asked for permission from the author [here](https://github.com/Pro/mdnsd/issues/17] if he agrees I will start development on the recipe.

Pro replied:
I agree :+1: Thanks for your effort!

madebr replied:
Just asking, isn't https://github.com/troglobit/mdnsd the upstream mdnsd?

Pro replied:
Yes, it was the original upstream repo. But at the time I integrated mDNS into open62541, that upstream repo was not maintained anymore. In my fork I fixed many memory issued and made it compilable for windows.
I guess the upstream repo is therefore currently not compatible with my fork.
At some point I need to merge these two again as it seems the upstream is active again.

saukijan replied:
first draft of [mdnsd](https://github.com/conan-io/conan-center-index/pull/2009) has been published as a PR! 

I am not 100% sure if the [test_package.cpp](https://github.com/conan-io/conan-center-index/blob/edf773c666424a22bcda9745a64eea9a835dcbb7/recipes/mdnsd/all/test_package/test_package.cpp) is the best implementation since I am not familiar with this library, if you have any ideas on how to improve it, please let me know!

Pro replied:
Unfortunately, my initial efforts to contribute all the necessary changes back to upstream were not really handled in a good way by the maintainer:

troglobit/mdnsd#1

Windows support was ignored at all.

Therefore I continued in my own fork. Checked yesterday how much effort it would be to create a new PR, but the fork already diverged too much.

madebr commented on recipes/open62541/all/conanfile.py:
```diff
@@ -0,0 +1,238 @@
+            raise ConanInvalidConfiguration(
+                "Clang compiler can not be used to build a static library")
+
+        cmake = CMake(self)
```
Please cache cmake object to avoid reconfiguration in package after build

saukijan replied:
Resolved in the next commit, just running local tests

SSE4 commented on recipes/open62541/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 2.8.12)
```
```suggestion
cmake_minimum_required(VERSION 3.1)
```
to fix a hook error
